src/Microsoft.DotNet.XHarness.CLI/Commands/WASM/Browser/WasmBrowserTestRunner.cs

```diff
@@ -232,14 +227,29 @@ private void RunSeleniumLogMessagePump(IWebDriver driver, CancellationToken toke
             }
         }
 
-        private string BuildUrl(string webServerAddr)
+        private string BuildUrl(ServerURLs serverURLs)
         {
-            var uriBuilder = new UriBuilder($"{webServerAddr}/{_arguments.HTMLFile}");
+            var uriBuilder = new UriBuilder($"{serverURLs.Http}/{_arguments.HTMLFile}");
             var sb = new StringBuilder();
 
             if (_arguments.DebuggerPort != null)
                 sb.Append($"arg=--debug");
 
+            if (_arguments.SetWebServerEnvironmentVariables)
+            {
+                var hostAndPort = serverURLs.Http.Substring(serverURLs.Http.LastIndexOf('/') + 1);
+                var hostAndPortSecure = serverURLs.Https.Substring(serverURLs.Https.LastIndexOf('/') + 1);
+
+                if (sb.Length > 0)
+                    sb.Append('&');
+                // see runtime\src\libraries\Common\tests\System\Net\Configuration.WebSockets.cs
+                // see runtime\src\libraries\Common\tests\System\Net\Configuration.Http.cs
+                sb.Append($"arg={HttpUtility.UrlEncode($"--setenv=DOTNET_TEST_WEBSOCKETHOST={hostAndPort}")}");
+                sb.Append($"&arg={HttpUtility.UrlEncode($"--setenv=DOTNET_TEST_SECUREWEBSOCKETHOST={hostAndPortSecure}")}");
+                sb.Append($"&arg={HttpUtility.UrlEncode($"--setenv=DOTNET_TEST_HTTPHOST={hostAndPort}")}");
+                sb.Append($"&arg={HttpUtility.UrlEncode($"--setenv=DOTNET_TEST_SECUREHTTPHOST={hostAndPortSecure}")}");
+            }
```

Comment:
Moving the list of variables to set into runtime repo makes good sense, thanks! 
I done it as two list (of http and https) variable names instead, because I don't want the web server middleware to have this integration responsibility.

Reply:
Also, the middleware doesn't know the address and port. And the URLencode is also complicating this.

Reply:
> Also, the middleware doesn't know the address and port. And the URLencode is also complicating this.

That's why I was suggesting having a method on that middleware type, to which you can pass the urls, and get back the `--setenv=DOTNET_TEST_WEBSOCKETHOST=http://...` string, which you can then `URLEncode` here.